### PR TITLE
More error handling in sway-lsp

### PR DESCRIPTION
**Pull request by sdankel**
- propagating errors instead of returning None in formatting.rs and `format_text`
- Move errors back into DirectoryError enum
- renaming `get_format_text_edits` to `get_page_text_edit` to be more clear
- flattening some instances of `None => None` using `and_then` for readability

### CODE REVIEW

File: sway-lsp/src/core/session.rs [hunk 1]
```diff
@@ -142,14 +142,10 @@ impl Session {
...
     pub fn declared_token_ident(&self, token: &Token) -> Option<Ident> {
-        // Look up the tokens TypeId
-        match &token.type_def {
-            Some(type_def) => match type_def {
-                TypeDefinition::TypeId(type_id) => utils::token::ident_of_type_id(type_id),
-                TypeDefinition::Ident(ident) => Some(ident.clone()),
-            },
-            None => None,
-        }
+        token.type_def.as_ref().and_then(|type_def| match type_def {
```
Comment (JoshuaBatty) [inline sway-lsp/src/core/session.rs:145]: lovely.

File: sway-lsp/src/core/session.rs [hunk 2]
```diff
@@ -73,7 +73,7 @@ impl Session {
 
         self.sync.watch_and_sync_manifest();
 
-        self.sync.manifest_dir()
+        self.sync.manifest_dir().map_err(Into::into)
```
Comment (JoshuaBatty) [inline sway-lsp/src/core/session.rs:76]: Easy, I'm fine with this error conversion happening at this spot.

**Review by JoshuaBatty — DISMISSED** [review]
Excellent, thanks @sdankel ! much nicer :)

**Review by kayagokalp — DISMISSED** [review]
utACK

File: sway-lsp/src/utils/sync.rs [hunk 3]
```diff
@@ -58,26 +58,26 @@ impl SyncWorkspace {
...
+            .ok_or(DirectoryError::ManifestDirNotFound)?;
 
         // extract the project name from the path
         let project_name = manifest_dir
             .file_name()
             .and_then(|name| name.to_str())
-            .ok_or(LanguageServerError::CantExtractProjectName {
+            .ok_or(DirectoryError::CantExtractProjectName {
                 dir: manifest_dir.to_string_lossy().to_string(),
             })?;
```
[the remark this answers is not in the document]
Comment (sdankel) [reply sway-lsp/src/utils/sync.rs:69]: What do you mean?

[the remark this answers is not in the document]
Comment (sdankel) [reply sway-lsp/src/utils/sync.rs:69]: @Centril Can you please clarify what you meant here?

File: sway-lsp/src/core/session.rs [hunk 4]
```diff
@@ -362,13 +356,17 @@ impl Session {
...
-            let mut formatter = Formatter::default();
-            get_format_text_edits(Arc::from(document.get_text()), &mut formatter)
-        } else {
-            None
-        }
+    pub fn format_text(&self, url: &Url) -> Result<Vec<TextEdit>, LanguageServerError> {
+        let document = self
+            .documents
+            .get(url.path())
+            .ok_or(DocumentError::DocumentNotFound {
```
[the remark this answers is not in the document]
Comment (sdankel) [reply sway-lsp/src/core/session.rs:363]: What do you mean? What's the functional difference here? @Centril

File: sway-lsp/src/core/session.rs [hunk 5]
```diff
@@ -336,16 +332,14 @@ impl Session {
             .and_then(|(_, token)| self.declared_token_ident(&token))
             .and_then(|decl_ident| {
                 let range = utils::common::get_range_from_span(&decl_ident.span());
-                match decl_ident.span().path() {
-                    Some(path) => match Url::from_file_path(path.as_ref()) {
-                        Ok(url) => self
-                            .sync
+                decl_ident.span().path().and_then(|path| {
+                    // We use ok() here because we don't care about propagating the error from from_file_path
+                    Url::from_file_path(path.as_ref()).ok().and_then(|url| {
```
[the remark this answers is not in the document]
Comment (sdankel) [reply sway-lsp/src/core/session.rs:337]: I'm not sure what you mean. How would I use `?` here?

File: sway-lsp/src/utils/sync.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (otrho) [reply sway-lsp/src/utils/sync.rs:69]: I think he's suggesting using `.ok_or_else(|| ...)` again as suggested above.

[the remark this answers is not in the document]
Comment (sdankel) [reply sway-lsp/src/utils/sync.rs:69]: Oh, I think you're saying to use `ok_or_else` instead

File: sway-lsp/src/capabilities/formatting.rs [hunk 6]
```diff
@@ -2,22 +2,25 @@ use std::sync::Arc;
...
+    let page_edit = TextEdit {
+        range: Range::new(Position::new(0, 0), Position::new(line_end as u32, 0)),
+        new_text: formatted_code,
+    };
 
-            Some(vec![main_edit])
-        }
-        _ => None,
-    }
+    Ok(page_edit)
```
[the remark this answers is not in the document]
Comment (sdankel) [reply sway-lsp/src/capabilities/formatting.rs:25]: Done.

**Review by JoshuaBatty — APPROVED** [review]

**Pull request merged**